File: backend/app/schemas/audit.py
"""
Pydantic schemas for the Audit Log / Activity resource — Phase 5.

Response models only — there are no create/update request models because
audit records are written exclusively by the service layer, never by API
consumers directly.

Security notes:
  - password_hash is never included.
  - JWT tokens are never included.
  - OTP values are never included.
  - SMTP or database credentials are never included.
"""

from datetime import datetime

from pydantic import BaseModel

from app.models.audit_log import AuditAction


# --------------------------------------------------------------------------- #
# Embedded sub-schemas                                                         #
# --------------------------------------------------------------------------- #

class AuditActorBrief(BaseModel):
    """Minimal user information embedded in audit log responses.

    Intentionally excludes password_hash, JWT, OTP, and all secrets.
    """

    id: int
    full_name: str
    email: str
    role: str

    model_config = {"from_attributes": True}


# --------------------------------------------------------------------------- #
# Response schemas                                                             #
# --------------------------------------------------------------------------- #

class AuditLogResponse(BaseModel):
    """Complete representation of a single audit record.

    old_values and new_values contain only the changed fields (diff).
    They will be None when not applicable (e.g. AUTH_LOGIN).
    """

    id: int
    user_id: int | None
    actor: AuditActorBrief | None
    action: AuditAction
    entity_type: str
    entity_id: int | None
    entity_key: str | None
    description: str
    old_values: dict | None
    new_values: dict | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogListResponse(BaseModel):
    """Paginated list of audit records, newest first."""

    items: list[AuditLogResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/audit.ts
import type { UserRole } from './auth';

export type AuditAction =
  | 'ISSUE_CREATED'
  | 'ISSUE_UPDATED'
  | 'ISSUE_ASSIGNED'
  | 'ISSUE_STATUS_CHANGED'
  | 'ISSUE_RESOLVED'
  | 'ISSUE_REOPENED'
  | 'PROJECT_CREATED'
  | 'PROJECT_UPDATED'
  | 'PROJECT_DEACTIVATED'
  | 'AUTH_LOGIN'
  | 'AUTH_LOGOUT'
  | 'USER_UPDATED'
  | 'USER_ACTIVATED'
  | 'USER_DEACTIVATED'
  | 'USER_ROLE_CHANGED'
  | 'COMMENT_CREATED'
  | 'COMMENT_UPDATED'
  | 'COMMENT_DELETED'
  | 'ATTACHMENT_UPLOADED'
  | 'ATTACHMENT_DELETED';

export interface AuditActorBrief {
  id: number;
  full_name: string;
  email: string;
  role: UserRole;
}

export interface AuditLogItem {
  id: number;
  user_id: number | null;
  actor: AuditActorBrief | null;
  action: AuditAction;
  entity_type: string;
  entity_id: number | null;
  entity_key: string | null;
  description: string;
  old_values: Record<string, any> | null;
  new_values: Record<string, any> | null;
  created_at: string;
}

export interface AuditLogListResponse {
  items: AuditLogItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
